helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java

```diff
@@ -1307,6 +1315,11 @@ public void shutdown() throws InterruptedException {
     logger.info("Shutting down {} pipeline", Pipeline.Type.TASK.name());
     shutdownPipeline(_taskEventThread, _taskEventQueue);
 
+    if (!_periodicalRebalanceExecutor
```

Comment:
Could supurious wake up can impact this?  I am not sure why we would like to wait here? Are we worrying about the partial task running? If it wakes up and return false, it will kill thread immediately. There is no difference as we kill it at first beginning.

Reply:
TFTR. 

Do you think it's better to do a _periodicalRebalanceExecutor.shutdownNow() at beginning and then check if no thread is running? Or just do _periodicalRebalanceExecutor.shutdownNow() at beginning is enough?

Reply:
I think it is fine. But it would be easier for the reader to understand if you put all the related logic together.

Reply:
TFTR. Updated.